src/snowflake/snowpark/mock/_plan.py

```diff
@@ -536,10 +536,17 @@ def cleanup_imports():
                 child, input_data, analyzer, expr_to_alias
             )
 
-        res = function_input.apply(lambda row: udf_handler(*row), axis=1)
-        res.sf_type = ColumnType(exp.datatype, exp.nullable)
-        res.name = quote_name(
-            f"{exp.udf_name}({', '.join(input_data.columns)})".upper()
+        # we do not use pd.apply here because pd.apply will auto infer dtype for the output column
+        # this will lead to NaN or None information loss, think about the following case of a udf definition:
+        #    def udf(x): return numpy.sqrt(x) if x is not None else None
+        # calling udf(-1) and udf(None), pd.apply will infer the column dtype to be int which returns NaT for both
+        # however, we want NaT for the former case and None for the latter case.
+        # using dtype object + function execution does not have the limitation
+        res = ColumnEmulator(
```

Comment:
In the future maybe we could call `fix_drift_between_column_sf_type_and_dtype` in methods like `set_sf_type`. And these code would look like:
```
res=input.apply(...)
res.set_sf_type(ColumnType(exp.datatype, exp.nullable))  # fixes the drift and removes NaT
```

Reply:
good point, added a comment